File: ios/sdk/swift/Connector.swift
//
//  Connector.swift
//  sdk
//
//  Created by mengxk on 2019/12/5.
//  Copyright © 2019 Elastos. All rights reserved.
//

import Foundation
import ContactSDK

open class Connector {
  private var mPeerNode: PeerNode?
  private var mServiceName: String
  
  private var mMessagListener: PeerNodeListener.MessageListener? = nil
  private var mDataListener: PeerNodeListener.DataListener? = nil
  
  public init(serviceName: String?) {
    mPeerNode = PeerNode.GetInstance()
    if (serviceName == nil || serviceName!.isEmpty) {
        mServiceName = PeerNode.CHAT_SERVICE_NAME;
    } else {
        mServiceName = serviceName!;
    }
  }
  
  deinit {
    removeMessageListener()
    removeDataListener()
  }
  
  public func setMessageListener(listener: PeerNodeListener.MessageListener) {
    if (mPeerNode == nil) {
      return
    }
    if (mMessagListener != nil) {
      removeMessageListener()
    }
    mPeerNode?.addMessageListener(serviceName: mServiceName, listener: listener)
    mMessagListener = listener
  }
  
  public func removeMessageListener() {
    if (mPeerNode == nil
    || mMessagListener == nil) {
      return
    }
    mPeerNode?.removeMessageListener(serviceName: mServiceName, listener: mMessagListener!)
    mMessagListener = nil
  }
  
  public func setDataListener(listener: PeerNodeListener.DataListener) {
    if (mPeerNode == nil) {
      return
    }
    if (mDataListener != nil) {
        removeDataListener()
    }
    mPeerNode?.addDataListener(serviceName: mServiceName, listener: listener)
    mDataListener = listener
  }
  
  public func removeDataListener() {
    if (mPeerNode == nil
    || mDataListener == nil) {
      return
    }
    mPeerNode?.removeDataListener(serviceName: mServiceName, listener: mDataListener!)
    mDataListener = nil
  }
  
  public func getUserInfo() -> Contact.UserInfo? {
      return mPeerNode?.getUserInfo()
  }
  
  public func addFriend(friendCode: String, summary: String) -> Int {
    if (mPeerNode == nil) {
      return -1
    }

    do {
      let data = [
        "serviceName": mServiceName,
        "content": summary,
      ]
      let encode = try JSONEncoder().encode(data)
      let val = String(data: encode, encoding: .utf8)!
      return mPeerNode!.addFriend(friendCode: friendCode, summary: val)
    } catch {
      print("make json failed\n")
    }
    
    return -1
  }

  public func removeFriend(friendCode: String) -> Int {
    return mPeerNode?.removeFriend(friendCode: friendCode) ?? -1
  }

  public func acceptFriend(friendCode: String) -> Int {
    return mPeerNode?.acceptFriend(friendCode: friendCode) ?? -1
  }
  
  public func setFriendInfo(friendCode: String, item: Contact.HumanInfo.Item, value: String) -> Int {
    return mPeerNode?.setFriendInfo(friendCode: friendCode, item: item, value: value) ?? -1
  }
  
  public func listFriendInfo() -> [ContactSDK.Contact.FriendInfo]? {
    return mPeerNode?.listFriendInfo() ?? nil
  }
  
  public func listFriendCode() -> [String] {
    return mPeerNode?.listFriendCode() ?? [String]()
  }
  
  public func getStatus() -> Contact.Status {
    return mPeerNode?.getStatus() ?? Contact.Status.Invalid
  }
  
  public func getFriendStatus(friendCode: String) -> Contact.Status {
    return mPeerNode?.getFriendStatus(friendCode: friendCode) ?? Contact.Status.Invalid
  }
  
  public func sendMessage(friendCode: String, channel: Contact.Channel, message: String) -> Int {
   if (mPeerNode == nil) {
     return -1
   }

    var data: String
    let isDidFriend = ContactSDK.Contact.IsDidFriend(friendCode: friendCode)
    if (isDidFriend) {
     do {
       let json = [
         "serviceName": mServiceName,
         "content": message,
       ]
       let encode = try JSONEncoder().encode(json)
       data = String(data: encode, encoding: .utf8)!
     } catch {
       print("make json failed\n")
       return -1
     }
    } else {
      data = message
    }
   
    return mPeerNode!.sendMessage(friendCode: friendCode, channel: channel,
                                  message: Contact.MakeTextMessage(text: data, cryptoAlgorithm: nil, memo: nil))
  }
  
  public func sendBinaryMessage(friendCode: String, channel: Contact.Channel, message: Data) -> Int {
    if (mPeerNode == nil) {
      return -1
    }

    var data: Data
    let isDidFriend = ContactSDK.Contact.IsDidFriend(friendCode: friendCode)
    if (isDidFriend) {
     do {
       let json = [
         "serviceName": mServiceName,
         "content": "binary",
       ]
       let encode = try JSONEncoder().encode(json)
       let str = String(data: encode, encoding: .utf8)!
       data = str.data(using: String.Encoding.utf8)!
       data.append(PeerNode.PROTOCOL_APPEND_DATA)
       data.append(message)
     } catch {
       print("make json failed\n")
       return -1
     }
    } else {
      data = message
    }

    return mPeerNode!.sendMessage(friendCode: friendCode, channel: channel,
                                  message: Contact.MakeBinaryMessage(data: data, cryptoAlgorithm: nil, memo: nil))
  }

}

File: android/sdk/src/main/java/app/elaphant/sdk/peernode/Connector.java
package app.elaphant.sdk.peernode;

import org.elastos.sdk.elephantwallet.contact.Contact;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public final class Connector {
    private static final String TAG = Connector.class.getName();

    private PeerNode mPeerNode = null;
    private String mServiceName;

    private PeerNodeListener.MessageListener mMessagListener = null;
    private PeerNodeListener.DataListener mDataListener = null;

    public Connector(String serviceName) {
        mPeerNode = PeerNode.getInstance();
        if (serviceName == null || serviceName.isEmpty()) {
            mServiceName = PeerNode.CHAT_SERVICE_NAME;
        } else {
            mServiceName = serviceName;
        }
    }

    public void finalize() {
        removeMessageListener();
        removeDataListener();
    }

    public void setMessageListener(PeerNodeListener.MessageListener listener) {
        if (mPeerNode == null) return;
        if (mMessagListener != null) {
            removeMessageListener();
        }
        mPeerNode.addMessageListener(mServiceName, listener);
        mMessagListener = listener;
    }

    public void removeMessageListener() {
        if (mPeerNode == null) return;
        mPeerNode.removeMessageListener(mServiceName, mMessagListener);
        mMessagListener = null;
    }

    public void setDataListener(PeerNodeListener.DataListener listener) {
        if (mPeerNode == null) return;
        if (mDataListener != null) {
            removeDataListener();
        }
        mPeerNode.addDataListener(mServiceName, listener);
        mDataListener = listener;
    }

    public void removeDataListener() {
        if (mPeerNode == null) return;
        if (mDataListener == null) return;
        mPeerNode.removeDataListener(mServiceName, mDataListener);
        mDataListener = null;
    }

    public Contact.UserInfo getUserInfo() {
        return mPeerNode.getUserInfo();
    }

    public int addFriend(String friendCode, String summary) {
        if (mPeerNode == null) return -1;

        JSONObject json = new JSONObject();
        try {
            json.put("serviceName", mServiceName);
            json.put("content", summary);
            return mPeerNode.addFriend(friendCode, json.toString());
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return -1;
    }

    public int removeFriend(String friendCode) {
        if (mPeerNode == null) return -1;
        return mPeerNode.removeFriend(friendCode);
    }

    public int acceptFriend(String friendCode) {
        if (mPeerNode == null) return -1;
        return mPeerNode.acceptFriend(friendCode);
    }

    public int setFriendInfo(String humanCode, Contact.HumanInfo.Item item, String value) {
        if (mPeerNode == null) return -1;
        return mPeerNode.setFriendInfo(humanCode, item, value);
    }

    public List<Contact.FriendInfo> listFriendInfo() {
        if (mPeerNode == null) return null;
        return mPeerNode.listFriendInfo();
    }

    public List<String> listFriendCode() {
        if (mPeerNode == null) return null;
        return mPeerNode.listFriendCode();
    }

    public Contact.Status getStatus() {
        if (mPeerNode == null) return Contact.Status.Invalid;
        return mPeerNode.getStatus();
    }

    public Contact.Status getFriendStatus(String friendCode) {
        if (mPeerNode == null) return Contact.Status.Invalid;
        return mPeerNode.getFriendStatus(friendCode);
    }

    public int sendMessage(String friendCode, Contact.Channel channel, String message) {
        if (mPeerNode == null) return -1;

        boolean isDidFriend = Contact.IsDidFriend(friendCode);

        String data;
        if (isDidFriend) {
            try {
                JSONObject json = new JSONObject();
                json.put("serviceName", mServiceName);
                json.put("content", message);
                data = json.toString();
            } catch (JSONException e) {
                e.printStackTrace();
                return -1;
            }
        } else {
            data = message;
        }

        return mPeerNode.sendMessage(friendCode, channel, Contact.MakeTextMessage(data, null));
    }

    public int sendBinaryMessage(String friendCode, Contact.Channel channel, byte[] binary) {
        if (mPeerNode == null) return -1;

        boolean isDidFriend = Contact.IsDidFriend(friendCode);

        byte[] data;
        if (isDidFriend) {
            try {
                JSONObject json = new JSONObject();
                json.put("serviceName", mServiceName);
                json.put("content", "binary");
                byte[] protocol = json.toString().getBytes();
                data = new byte[protocol.length + PeerNode.PROTOCOL_APPEND_DATA.length + binary.length];
                System.arraycopy(protocol, 0, data, 0, protocol.length);
                System.arraycopy(PeerNode.PROTOCOL_APPEND_DATA, 0, data, protocol.length, PeerNode.PROTOCOL_APPEND_DATA.length);
                System.arraycopy(binary, 0, data, protocol.length + PeerNode.PROTOCOL_APPEND_DATA.length, binary.length);
            } catch (JSONException e) {
                e.printStackTrace();
                return -1;
            }
        } else {
            data = binary;
        }

        return mPeerNode.sendMessage(friendCode, channel, Contact.MakeBinaryMessage(data, null));
    }
}
